### PR TITLE
capture AUT uncaught errors and properly route them to uncaught:exception handlers

### LINKED ISSUE #987: Cypress failing after uncaught:exception thrown from 3rd party, even thought 'uncaught:exception' handler is listening.
### Current behavior:

On the initial spec run, with a new browser, the exception is thrown from my application

```
TypeError: Cannot read property 'page_type' of undefined at getPageType
```

If I rerun the test, without closing the browser, the test passes and the error is not thrown.

I don't expect you to solve my applications errors. However, this does not occur in a local chrome window, might be useful.

The real issue is when the uncaught exception is thrown by my application it stops Cypress from executing the test and any other tests. However, the page still loads. (See the video attached)

If I am correct, Cypress should not stop with application errors with

```js
Cypress.on('uncaught:exception', (err, runnable) => {
    // returning false here prevents Cypress from
    // failing the test
    return false
})
```

in commands.js, which it is

### Desired behavior:
Cypress does not stop executing when the application throws an exception. 

### Additional Info (images, stack traces, etc)

https://youtu.be/m0O_P0alpyc

- Operating System: Mac
- Cypress Version: 1.1.2
- Browser Version: Chrome 62

### PR DESCRIPTION
**Pull request by kuceb**
- fix #987  
- [x] set top.onerror to same as autWindow.onerror to catch code in AUT that calls on `top` and throws in async code
- ~try/catch eval'd code in wrapped timer calls~ _its fine to let them bubble to top.onerror_
- [x] instead of immediately failing the test, allow errors in XHR.onload to be caught in `uncaught:exception` hook (thus ignored)

### Description of change

- fixed bug where certain types of uncaught errors originating from the AUT would not route through the `uncaught:exception` handler, and thus could not be ignored. These errors will also now correctly indicate that they came from application code.

### How has the user experience changed?

Code that throws from the AUT originating from XHR callback handlers or `setTimeouts`, etc now route correctly through Cypress's `uncaught:exception` handler, allowing users to bypass these errors.

Errors will now correctly indicate that they came from the AUT.

**Before:**

Screenshot 2019-10-30 17 49 29

**After:**

Screenshot 2019-10-30 17 51 41

### Pre-merge Tasks

- [x] Have tests been added/updated for the changes in this PR?
- [x] Has the original issue been tagged with a release in ZenHub?

### CODE REVIEW

File: packages/driver/test/cypress/integration/e2e/uncaught_errors_spec.coffee [hunk 1]
```diff
@@ -110,3 +110,16 @@ describe "uncaught errors", ->
...
 
       .contains("visit").click()
+
+  it 'global onerror', (done) ->
+    cy.once 'uncaught:exception', ->
+      setTimeout ->
+        done()
+      false
+
+    # cy.visit('https://www.nba.com/membership/user/login')
```
Comment (bahmutov) [inline packages/driver/test/cypress/integration/e2e/uncaught_errors_spec.coffee]: haha, nice comment

Comment (Bkucera) [reply packages/driver/test/cypress/integration/e2e/uncaught_errors_spec.coffee]: ok ok.

File: packages/driver/src/cy/commands/xhr.coffee [hunk 2]
```diff
@@ -180,8 +180,7 @@ startXhrServer = (cy, state, config) ->
       if log = logs[xhr.id]
         log.snapshot("error").error(err)
 
-      if r = state("reject")
-        r(err)
+      throw err
```
Comment (brian-mann) [inline packages/driver/src/cy/commands/xhr.coffee]: leave a note here indicating why this is necessary

File: packages/driver/src/cypress/cy.coffee [hunk 3]
```diff
@@ -56,6 +56,19 @@ create = (specWindow, Cypress, Cookies, state, config, log) ->
   stopped = false
   commandFns = {}
 
+  top.onerror = ->
```
Comment (brian-mann) [inline packages/driver/src/cypress/cy.coffee]: assign this and reuse it on the `get`, as opposed to pulling it off of `getOwnPropertyDescriptor` again

File: packages/driver/test/cypress/integration/commands/xhr_spec.coffee [hunk 4]
```diff
@@ -2,6 +2,11 @@ _ = Cypress._
 $ = Cypress.$
 Promise = Cypress.Promise
 
+uncaughtExceptionHookCalled = false
+
+Cypress.on 'uncaught:exception', ->
+  uncaughtExceptionHookCalled = true
+
```
Comment (brian-mann) [inline packages/driver/test/cypress/integration/commands/xhr_spec.coffee]: make this local to the function and use `cy.stub()`

**Review by brian-mann — CHANGES_REQUESTED** [review]

**Review by brian-mann — DISMISSED** [review]

**Pull request merged**